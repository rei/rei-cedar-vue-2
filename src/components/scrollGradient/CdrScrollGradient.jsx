import clsx from 'clsx';
import debounce from 'lodash-es/debounce';
import { CdrColorBackgroundPrimary } from '@rei/cdr-tokens/dist/js/cdr-tokens.esm';
import style from './styles/CdrScrollGradient.scss';
import BuildClass from '../../mixins/buildClass';

export default {
  name: 'CdrScrollGradient',
  mixins: [BuildClass],
  props: {
    // TODO: use modifier?
    horizontal: {
      type: Boolean,
      default: false,
    },
    vertical: {
      type: Boolean,
      default: false,
    },
    backgroundColor: {
      type: String,
      default: CdrColorBackgroundPrimary,
    },
    // height: {
    //   type: String,
    // },
    // width: {
    //   type: String,
    // },
  },
  data() {
    return {
      style,
      overflowLeft: false,
      overflowRight: false,
      overflowTop: false,
      overflowBottom: false,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-scroll-gradient';
    },
    styleClass() {
      const styles = [];

      if (this.horizontal) {
        styles.push(this.modifyClassName(this.baseClass, 'horizontal'));
      }

      if (this.vertical) {
        styles.push(this.modifyClassName(this.baseClass, 'vertical'));
      }

      return styles.join(' ');
    },
    gradientLeftStyle() {
      return this.createGradientStyle('left');
    },
    gradientRightStyle() {
      return this.createGradientStyle('right');
    },
    gradientTopStyle() {
      return this.createGradientStyle('top');
    },
    gradientBottomStyle() {
      return this.createGradientStyle('bottom');
    },
  },
  mounted() {
    this.$nextTick(() => {
      this.calculateOverflow();
    });
    // Check for header overflow on window resize for gradient behavior.
    window.addEventListener('resize', debounce(() => {
      this.calculateOverflow();
    }, 500));
    // Check for header overflow on widow resize for gradient behavior.
    this.$refs.cdrScrollGradientContent.addEventListener('scroll', debounce(() => {
      this.calculateOverflow();
    }, 50));
  },
  methods: {
    calculateOverflow() {
      if (this.vertical) {
        const contentHeight = this.$refs.cdrScrollGradientContent.children[0].offsetHeight;
        const containerHeight = this.$refs.cdrScrollGradientContainer.offsetHeight;
        if (contentHeight > containerHeight) {
          const scrollY = this.$refs.cdrScrollGradientContent.scrollTop;
          this.overflowTop = scrollY > 1;
          this.overflowBottom = (scrollY + 1) < (contentHeight - containerHeight);
        } else {
          this.overflowTop = false;
          this.overflowBottom = false;
        }
      }

      if (this.horizontal) {
        const contentWidth = this.$refs.cdrScrollGradientContent.children[0].offsetWidth;
        const containerWidth = this.$refs.cdrScrollGradientContainer.offsetWidth;
        if (contentWidth > containerWidth) {
          const scrollX = this.$refs.cdrScrollGradientContent.scrollLeft;
          this.overflowLeft = scrollX > 1;
          this.overflowRight = (scrollX + 1) < (contentWidth - containerWidth);
        } else {
          this.overflowLeft = false;
          this.overflowRight = false;
        }
      }
    },
    createGradientStyle(dir) {
      const bg = `linear-gradient(to ${dir}, rgba(255, 255, 255, 0), ${this.backgroundColor})`;
      return {
        background: bg,
      };
    },
  },
  render() {
    return (
      <div class={this.style[this.baseClass]} ref="cdrScrollGradientContainer">
        <div class={clsx(
          this.style['cdr-scroll-gradient--overlay'],
          this.style['cdr-scroll-gradient--overlay-left'],
          this.overflowLeft ? this.style['cdr-scroll-gradient--overlay-active'] : '',
        )}
          style={this.gradientLeftStyle}
        ></div>

        <div class={clsx(
          this.style['cdr-scroll-gradient--overlay'],
          this.style['cdr-scroll-gradient--overlay-top'],
          this.overflowTop ? this.style['cdr-scroll-gradient--overlay-active'] : '',
        )}
          style={this.gradientTopStyle}
        ></div>


        <div class={clsx(this.style['cdr-scroll-gradient--content'], this.styleClass)}
          ref="cdrScrollGradientContent">
          { this.$slots.default }
        </div>


        <div class={clsx(
          this.style['cdr-scroll-gradient--overlay'],
          this.style['cdr-scroll-gradient--overlay-right'],
          this.overflowRight ? this.style['cdr-scroll-gradient--overlay-active'] : '',
        )}
          style={this.gradientRightStyle}
        ></div>


        <div class={clsx(
          this.style['cdr-scroll-gradient--overlay'],
          this.style['cdr-scroll-gradient--overlay-bottom'],
          this.overflowBottom ? this.style['cdr-scroll-gradient--overlay-active'] : '',
        )}
          style={this.gradientBottomStyle}
        ></div>
      </div>
    );
  },
};
