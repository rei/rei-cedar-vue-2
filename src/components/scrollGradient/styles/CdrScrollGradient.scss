@import '../../../css/settings/index.scss';

.cdr-scroll-gradient {
  position: relative;
  height: inherit;
  width: inherit;


  &--content {
    height: inherit;
    width: inherit;
  }

  &--overlay {
    transition: opacity $cdr-duration-2-x ease;
    -webkit-transition: opacity $cdr-duration-2-x ease;
    position: absolute;
    z-index: 100;

    pointer-events: none;
    opacity: 0;
// TODO: DE-DUPE styles into vert/horiz
    &-left {
      top: 0;
      bottom: 0;
      left: 0;

      width: $cdr-space-three-x;
    }

    &-right {
      top: 0;
      bottom: 0;
      right: 0;

      width: $cdr-space-three-x;
    }

    &-top {
      top: 0;
      left: 0;
      right: 0;

      height: $cdr-space-three-x;
    }

    &-bottom {
      left: 0;
      bottom: 0;
      right: 0;

      height: $cdr-space-three-x;
    }

    &-active {
      opacity: 1;
    }
  }

  &--horizontal {
    overflow-x: auto;
    // overflow-y: hidden;
    // white-space: nowrap;
  }

  &--vertical {
    overflow-y: auto;
    // overflow-x: hidden
  }

}
