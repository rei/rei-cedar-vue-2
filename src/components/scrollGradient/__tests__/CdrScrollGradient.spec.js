import { shallowMount } from '@vue/test-utils';
import CdrScrollGradient from 'componentdir/scrollGradient/CdrScrollGradient';

describe('CdrScrollGradient', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(CdrScrollGradient);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('has a failing test by default so you remember to do them', () => {
    expect(false).toBe(true);
  });
});
