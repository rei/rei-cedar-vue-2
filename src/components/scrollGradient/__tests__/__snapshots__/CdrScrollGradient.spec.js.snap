// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrScrollGradient matches snapshot 1`] = `
<div
  class="cdr-scroll-gradient"
>
  <div
    class="cdr-scroll-gradient--container"
  />
</div>
`;
